data_structures/binary_tree/is_bst.py

```diff
@@ -0,0 +1,117 @@
+"""
+Author  : Alexander Pantyukhin
+Date    : November 2, 2022
+
+Task:
+Given the root of a binary tree, determine if it is a valid binary search
+tree (BST).
+
+A valid BST is defined as follows:
+
+- The left subtree of a node contains only nodes with keys less than the node's key.
+- The right subtree of a node contains only nodes with keys greater than the node's key.
+- Both the left and right subtrees must also be binary search trees.
+
+Implementation notes:
+Depth-first search approach.
+
+leetcode: https://leetcode.com/problems/validate-binary-search-tree/
+
+Let n is the number of nodes in tree
+Runtime: O(n)
+Space: O(1)
```

Comment:
How do we _prove_ these statements?

Reply:
About runtime: We visit each node of tree only one time. So it's the linear time alg based on the number of nodes. About space: we don't add new data strcututre inside the function.